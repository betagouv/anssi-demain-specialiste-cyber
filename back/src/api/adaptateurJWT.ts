import jwt, { JwtPayload } from 'jsonwebtoken';

export interface AdaptateurJWT {
  genereToken(donnees: Record<string, unknown>): string;
  decode(token: string): JwtPayload;
}

export const adaptateurJWT: AdaptateurJWT = {
  genereToken: (donnees: Record<string, unknown>) =>
    jwt.sign(donnees, process.env.SECRET_JWT || '', { expiresIn: '1h' }),
  decode: (token: string) =>
    jwt.verify(token, process.env.SECRET_JWT || '') as JwtPayload,
};
