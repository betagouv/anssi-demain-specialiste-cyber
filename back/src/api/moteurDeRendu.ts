import express from 'express';
import fs from 'fs';
import { join } from 'path';

export interface MoteurDeRendu {
  rends: (reponse: express.Response, vue: string, options?: object) => void;
}

type FournisseurDeChemins = () => { dsc: string; style: string };

const lisLesCheminsDesFichiersGeneresParLeFront: FournisseurDeChemins = () => {
  const cheminVersLeFichierManifeste = join(
    process.cwd(),
    '../front/dist/.vite/manifest.json',
  );
  const manifeste = JSON.parse(
    fs.readFileSync(cheminVersLeFichierManifeste, 'utf-8'),
  );
  return {
    dsc: manifeste['src/index.ts'].file,
    style: manifeste['style.css'].file,
  };
};

export const moteurDeRenduExpress = (
  fournisseurDeChemins: FournisseurDeChemins = lisLesCheminsDesFichiersGeneresParLeFront,
): MoteurDeRendu => {
  return {
    rends(reponse, vue, options) {
      const optionsAvecManifesteEtNonce = {
        ...options,
        ...fournisseurDeChemins(),
        nonce: reponse.locals.nonce,
      };
      reponse.render(vue, optionsAvecManifesteEtNonce);
    },
  };
};
