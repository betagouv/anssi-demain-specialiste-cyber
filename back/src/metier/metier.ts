export type Metier = {
  id: number;
  titre: string;
  description: string;
  missionPrincipale: string;
  postures: string[];
  formationsCibles: string[];
  preRequis: string[];
  remuneration: {
    junior: number;
    senior: number;
  };
  metiersProches: string[];
  liens: {
    illustration: string;
    dataemploi: string;
    metierscope: string;
    video: string;
  };
};
