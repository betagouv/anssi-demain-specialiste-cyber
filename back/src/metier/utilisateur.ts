interface InformationsCreationUtilisateur {
  email: string;
  prenom: string;
  nom: string;
  siretEntite: string;
  infolettreAcceptee: boolean;
}

export class Utilisateur {
  email: string;
  prenom: string;
  nom: string;
  infolettreAcceptee: boolean;
  siretEntite: string;

  constructor({
    email,
    prenom,
    nom,
    infolettreAcceptee,
    siretEntite,
  }: InformationsCreationUtilisateur) {
    this.email = email;
    this.prenom = prenom;
    this.nom = nom;
    this.infolettreAcceptee = infolettreAcceptee;
    this.siretEntite = siretEntite;
  }
}
