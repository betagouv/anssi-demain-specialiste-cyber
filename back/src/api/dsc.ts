import { creeServeurLab } from '@lab-anssi/lib';
import cookieParser from 'cookie-parser';
import cookieSession from 'cookie-session';
import express, { json } from 'express';
import { ConfigurationServeur } from './configurationServeur';
import { ressourcesApi } from './ressourcesApi';
import { ressourcesOidc } from './ressourcesOidc';
import { ressourcesPages } from './ressourcesPages';

export const creeServeur = (configurationServeur: ConfigurationServeur) => {
  const { serveurLab } = configurationServeur;
  const app = creeServeurLab(serveurLab);

  app.use(configurationServeur.middleware.ajouteLeNonceALaReponse);
  app.use(configurationServeur.middleware.positionneLesCsp);

  app.use(json());
  app.use(
    cookieSession({
      name: 'session',
      sameSite: true,
      secret: process.env.SECRET_COOKIE,
      secure: configurationServeur.adaptateurEnvironnement.nodeEnv() !== 'developpement',
      signed: configurationServeur.adaptateurEnvironnement.nodeEnv() !== 'developpement',
    }),
  );

  app.use(cookieParser());

  configurationServeur
    .recupereCheminsVersFichiersStatiques()
    .forEach((chemin) => {
      app.use('/', express.static(chemin));
    });

  app.use(configurationServeur.middleware.verifieModeMaintenance);
  app.use('/oidc', ressourcesOidc(configurationServeur));
  app.use('/api', ressourcesApi(configurationServeur));

  app.set('view engine', 'pug');
  app.set('views', './vues');
  app.use(ressourcesPages(configurationServeur));

  return app;
};
