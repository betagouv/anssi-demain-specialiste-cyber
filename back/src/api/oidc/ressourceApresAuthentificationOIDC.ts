import { Router } from 'express';
import { ConfigurationServeur } from '../dsc';

const ressourceApresAuthentificationOIDC = ({
  adaptateurOIDC,
  adaptateurJWT,
  entrepotUtilisateur,
  adaptateurHachage,
}: ConfigurationServeur) => {
  const routeur = Router();
  routeur.get('/', async (requete, reponse) => {
    if (!requete.cookies.AgentConnectInfo) {
      reponse.sendStatus(401);
      return;
    }

    try {
      const { accessToken, idToken } =
        await adaptateurOIDC.recupereJeton(requete);
      const informationsUtilisateur =
        await adaptateurOIDC.recupereInformationsUtilisateur(accessToken);
      const { email } = informationsUtilisateur;

      if (!(await entrepotUtilisateur.existe(adaptateurHachage.hache(email)))) {
        const token = adaptateurJWT.genereToken(informationsUtilisateur);
        reponse.redirect(`/creation-compte?token=${token}`);
        return;
      }

      requete.session = { ...requete.session, ...informationsUtilisateur };
      requete.session.token = adaptateurJWT.genereToken({ email });
      requete.session.AgentConnectIdToken = idToken;
      reponse.sendFile(`./pages/apres-authentification.html`, {
        root: '.',
      });
    } catch (e: unknown | Error) {
      // eslint-disable-next-line no-console
      console.error('Erreur apr√®s authentification', e);
      reponse.sendStatus(401);
    }
  });
  return routeur;
};

export { ressourceApresAuthentificationOIDC };
