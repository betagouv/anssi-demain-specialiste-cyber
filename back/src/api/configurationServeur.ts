import { ConfigurationServeurLab } from '@lab-anssi/lib';
import { BusEvenements } from '../bus/busEvenements';
import { AdaptateurEnvironnement } from '../infra/adaptateurEnvironnement';
import { AdaptateurHachage } from '../infra/adaptateurHachage';
import { AdaptateurJournal } from '../infra/adaptateurJournal';
import { AdaptateurRechercheEntreprise } from '../infra/adaptateurRechercheEntreprise';
import { RecupereCheminVersFichiersStatiques } from '../infra/recupereCheminVersFichiersStatiques';
import { EntrepotJeux } from '../metier/entrepotJeux';
import { EntrepotRessourcesCyber } from '../metier/entrepotRessourcesCyber';
import { EntrepotUtilisateur } from '../metier/entrepotUtilisateur';
import { AdaptateurJWT } from './adaptateurJWT';
import { Middleware } from './middleware';
import { MoteurDeRendu } from './moteurDeRendu';
import { AdaptateurOIDC } from './oidc/adaptateurOIDC';
import { AdaptateurTeleversement } from '../infra/adaptateurTeleversement';

export interface ConfigurationServeur {
  adaptateurEnvironnement: AdaptateurEnvironnement;
  serveurLab: ConfigurationServeurLab;
  entrepotRessourcesCyber: EntrepotRessourcesCyber;
  adaptateurOIDC: AdaptateurOIDC;
  adaptateurJWT: AdaptateurJWT;
  adaptateurRechercheEntreprise: AdaptateurRechercheEntreprise;
  entrepotUtilisateur: EntrepotUtilisateur;
  adaptateurHachage: AdaptateurHachage;
  recupereCheminsVersFichiersStatiques: RecupereCheminVersFichiersStatiques;
  middleware: Middleware;
  moteurDeRendu: MoteurDeRendu;
  busEvenements: BusEvenements;
  entrepotJeux: EntrepotJeux;
  adaptateurJournal: AdaptateurJournal;
  adaptateurTeleversement: AdaptateurTeleversement;
}

export type ConfigurationServeurSansMiddleware = Omit<
  ConfigurationServeur,
  'middleware'
>;
