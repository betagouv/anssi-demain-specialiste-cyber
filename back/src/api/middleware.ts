import { HttpStatusCode } from 'axios';
import { NextFunction, Request, Response } from 'express';
import { randomBytes } from 'node:crypto';
import z from 'zod';
import { ConfigurationServeurSansMiddleware } from './configurationServeur';
import { AdaptateurHachage } from '../infra/adaptateurHachage';
import { EntrepotUtilisateur } from '../metier/entrepotUtilisateur';
import { Utilisateur } from '../metier/utilisateur';

type FonctionMiddleware<TBody> = (
  requete: Request<unknown, unknown, TBody, unknown, never>,
  reponse: Response,
  suite: NextFunction,
) => Promise<void>;

export type Middleware = {
  verifieModeMaintenance: FonctionMiddleware<unknown>;
  valideLaCoherenceDuCorps: <
    TZod extends z.ZodType,
    TBody extends z.infer<TZod>,
  >(
    object: TZod,
  ) => FonctionMiddleware<TBody>;
  verifieJWTNavigation: FonctionMiddleware<unknown>;
  ajouteUtilisateurARequete(
    entrepotUtilisateur: EntrepotUtilisateur,
    adaptateurHachage: AdaptateurHachage,
  ): FonctionMiddleware<unknown>;
};

export type RequeteNonTypee = Request<
  unknown,
  unknown,
  unknown,
  unknown,
  never
>;

export const fabriqueMiddleware = ({
  adaptateurEnvironnement,
  adaptateurJWT,
}: ConfigurationServeurSansMiddleware): Middleware => {
  const verifieModeMaintenance = async (
    _requete: RequeteNonTypee,
    reponse: Response,
    suite: NextFunction,
  ) => {
    if (adaptateurEnvironnement.maintenance().actif()) {
      const nonce = randomBytes(24).toString('base64');
      reponse
        .status(HttpStatusCode.ServiceUnavailable)
        .set('Content-Type', 'text/html')
        .render('maintenance', { nonce });
    } else {
      suite();
    }
  };

  const valideLaCoherenceDuCorps =
    <TZod extends z.ZodType, TBody extends z.infer<TZod>>(objet: TZod) =>
    async (
      requete: Request<unknown, unknown, TBody, unknown, never>,
      reponse: Response,
      suite: NextFunction,
    ) => {
      const resultat = objet.safeParse(requete.body);
      if (!resultat.success) {
        reponse.status(400).json({ erreur: resultat.error.issues[0].message });
      } else {
        suite();
      }
    };

  const verifieJWTNavigation = async (
    requete: RequeteNonTypee,
    reponse: Response,
    suite: NextFunction,
  ) => {
    if (!requete.session?.token) {
      return reponse.redirect('/connexion');
    }
    try {
      adaptateurJWT.decode(requete.session.token);
      suite();
    } catch {
      reponse.redirect('/connexion');
    }
  };

  const ajouteUtilisateurARequete =
    (
      entrepotUtilisateur: EntrepotUtilisateur,
      adaptateurHachage: AdaptateurHachage,
    ) =>
    async (
      requete: RequeteNonTypee & { utilisateur?: Utilisateur | undefined },
      reponse: Response,
      suite: NextFunction,
    ) => {
      try {
        const { email } = adaptateurJWT.decode(requete.session?.token);
        requete.utilisateur = email
          ? await entrepotUtilisateur.parEmailHache(
              adaptateurHachage.hache(email),
            )
          : undefined;
        suite();
      } catch {
        reponse.sendStatus(500);
      }
    };

  return {
    verifieModeMaintenance,
    valideLaCoherenceDuCorps,
    verifieJWTNavigation,
    ajouteUtilisateurARequete,
  };
};
