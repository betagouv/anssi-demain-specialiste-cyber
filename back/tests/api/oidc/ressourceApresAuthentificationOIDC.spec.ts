import { Express } from 'express';
import request from 'supertest';
import { beforeEach, describe, expect, it } from 'vitest';
import { AdaptateurJWT } from '../../../src/api/adaptateurJWT';
import { ConfigurationServeur, creeServeur } from '../../../src/api/dsc';
import { AdaptateurOIDC } from '../../../src/api/oidc/adaptateurOIDC';
import { Utilisateur } from '../../../src/metier/utilisateur';
import { EntrepotUtilisateurMemoire } from '../../infra/entrepotUtilisateurMemoire';
import { decodeSessionDuCookie } from '../cookie';
import {
  configurationDeTestDuServeur,
  fauxAdaptateurJWT,
  fauxAdaptateurOIDC,
} from '../fauxObjets';

describe('La ressource apres authentification OIDC', () => {
  describe('quand on fait un GET sur /oidc/apres-authentification', () => {
    let serveur: Express;
    let adaptateurOIDC: AdaptateurOIDC;
    let adaptateurJWT: AdaptateurJWT;
    let entrepotUtilisateur: EntrepotUtilisateurMemoire;

    beforeEach(() => {
      adaptateurOIDC = { ...fauxAdaptateurOIDC };
      adaptateurJWT = fauxAdaptateurJWT;
      entrepotUtilisateur = new EntrepotUtilisateurMemoire();
      const configurationServeur: ConfigurationServeur = {
        ...configurationDeTestDuServeur(),
        adaptateurOIDC,
        adaptateurJWT,
        entrepotUtilisateur,
      };
      serveur = creeServeur(configurationServeur);
    });

    const requeteGet = () =>
      request(serveur)
        .get('/oidc/apres-authentification')
        .set('Cookie', ['AgentConnectInfo={}']);

    describe("si l'utilisateur est connu", () => {
      beforeEach(async () => {
        const jeanneDupont = new Utilisateur({
          email: 'jeanne.dupont',
          infolettreAcceptee: true,
          prenom: '',
          nom: '',
          siretEntite: '',
        });
        await entrepotUtilisateur.ajoute(jeanneDupont);

        adaptateurOIDC.recupereInformationsUtilisateur = async (_) => ({
          prenom: 'Jeanne',
          nom: 'Dupont',
          email: 'jeanne.dupont',
          siret: '1234',
        });
      });

      it('reçoit 200', async () => {
        const reponse = await requeteGet();

        expect(reponse.status).toBe(200);
      });

      it("ajoute les informations de l'utilisateur à la session", async () => {
        adaptateurOIDC.recupereJeton = async () => {
          return { idToken: 'xx', accessToken: 'y' };
        };
        adaptateurOIDC.recupereInformationsUtilisateur = async (
          accessToken
        ) => {
          if (accessToken === 'y') {
            return {
              prenom: 'Jeanne',
              nom: 'Dupont',
              email: 'jeanne.dupont',
              siret: '1234',
            };
          }
          throw new Error('Aurait du être appelé avec le bon access token');
        };

        const reponse = await requeteGet();

        const session = decodeSessionDuCookie(reponse, 0);
        expect(session).toBeDefined();
        expect(session.prenom).toBe('Jeanne');
        expect(session.nom).toBe('Dupont');
        expect(session.email).toBe('jeanne.dupont');
        expect(session.siret).toBe('1234');
      });

      it('ajoute un token JWT à la session', async () => {
        adaptateurJWT.genereToken = (donnees: Record<string, unknown>) =>
          `tokenJWT-${donnees.email}`;

        const reponse = await requeteGet();

        const session = decodeSessionDuCookie(reponse, 0);
        expect(session.token).toBe('tokenJWT-jeanne.dupont');
      });

      it('ajoute un tokenId AgentConnect à la session', async () => {
        adaptateurOIDC.recupereJeton = async () => {
          return { idToken: 'tokenAgentConnect', accessToken: 'y' };
        };

        const reponse = await requeteGet();

        const session = decodeSessionDuCookie(reponse, 0);
        expect(session.AgentConnectIdToken).toBe('tokenAgentConnect');
      });

      it('sert la page apres-authentification', async () => {
        const reponse = await requeteGet();

        expect(reponse.headers['content-type']).toEqual(
          'text/html; charset=utf-8'
        );
        expect(reponse.text).toMatchSnapshot();
      });
    });

    it("jette une erreur 401 si le cookie AgentConnectInfo n'est pas défini", async () => {
      const reponse = await requeteGet().set('Cookie', []);

      expect(reponse.status).toBe(401);
    });

    it('jette une erreur 401 si quoi que ce soit se passe mal', async () => {
      adaptateurOIDC.recupereJeton = async () => {
        throw new Error('mauvais state');
      };

      const reponse = await requeteGet();

      expect(reponse.status).toBe(401);
    });

    describe("si l'utilisateur est inconnu", () => {
      it('ajoute un token contenant les informations du nouvel utilisateur et redirige vers la page de création de compte', async () => {
        const reponse = await requeteGet();

        expect(reponse.status).toBe(302);
        expect(reponse.headers.location).toBe(
          '/creation-compte?token=tokenJWT-'
        );
      });
    });
  });
});
