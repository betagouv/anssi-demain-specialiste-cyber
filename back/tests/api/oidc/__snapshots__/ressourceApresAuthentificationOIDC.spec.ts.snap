// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`La ressource apres authentification OIDC > quand on fait un GET sur /oidc/apres-authentification > si l'utilisateur est connu > sert la page apres-authentification 1`] = `
"<!DOCTYPE html><html><head><script>document.addEventListener('DOMContentLoaded', () => {
  window.location = '/catalogue';
});</script></head></html>"
`;
